### DISCUSSION TITLE
Clarifications regarding behaviour when there are REKEY collisions

### DISCUSSION THREAD

**Discussion by amvarunn** [strongswan/strongswan#1234] [Q&A]
Hi, please help us in understanding the below doubts.

Attaching event dumps of interest.

 

 

```
======== Event start ========

Time: Tue Aug 16 16:32:40 2022

  child-rekey {

    ED_USR_IPR_1_0_0_2_65535_IPVX_192-168-180-15_192-168-180-16_3864 {

      uniqueid = 1

      version = 2

      state = ESTABLISHED

       ....

       ....

      tasks-active [

        CHILD_REKEY

      ]

      tasks-passive [

        CHILD_REKEY

      ]

      child-sas {

        ED_USR_IPR_1_0_0_2_65535_IPVX_192-168-180-15_192-168-180-16_3864 {

          old {

            name = ED_USR_IPR_1_0_0_2_65535_IPVX_192-168-180-15_192-168-180-16_3864

            uniqueid = 20

            reqid = 1

            state = REKEYED

            mode = TUNNEL

            protocol = ESP

            spi-in = c4b60749

            spi-out = cd8176fe

            .......

            .......

          }

          new {

            name = ED_USR_IPR_1_0_0_2_65535_IPVX_192-168-180-15_192-168-180-16_3864

            uniqueid = 22

            reqid = 1

            state = INSTALLED

            mode = TUNNEL

            protocol = ESP

            spi-in = c7a416d3

            spi-out = c06264c5

.....

.....

          }

        }

      }

    }

  }

======== Event start ========

  child-rekey {

    ED_USR_IPR_1_0_0_2_65535_IPVX_192-168-180-15_192-168-180-16_3864 {

      uniqueid = 1

      version = 2

      state = ESTABLISHED

....

.....

      tasks-active [

        CHILD_REKEY

      ]

      child-sas {

        ED_USR_IPR_1_0_0_2_65535_IPVX_192-168-180-15_192-168-180-16_3864 {

          old {

            name = ED_USR_IPR_1_0_0_2_65535_IPVX_192-168-180-15_192-168-180-16_3864

            uniqueid = 20

            reqid = 1

            state = REKEYED

            mode = TUNNEL

            protocol = ESP

            spi-in = c4b60749

            spi-out = cd8176fe

             ....

             ....

          }

          new {

            name = ED_USR_IPR_1_0_0_2_65535_IPVX_192-168-180-15_192-168-180-16_3864

            uniqueid = 21

            reqid = 1

            state = INSTALLED

            mode = TUNNEL

            protocol = ESP

            spi-in = c4a7d179

            spi-out = cb461b99

   ....

   ....

       }

        }

      }

    }

  }
```

 

 

In the above events, we see spi-out cd8176fe twice, If we were to match these fields and do an increment on "new" and decrement on "old", we would be decrementing twice for a single increment. 

Please confirm if this is the actual behaviour.

 

The second question is, in case of CHILD_REKEY collisions, we sometimes see CHILD_DOWN and other times we do not.

What is the actual behaviour in this case?
@tobiasbrunner , @strongX509 

Thank you

**Comment by tobiasbrunner** [comment]
Yes, that's a known issue.  If the rekey collision is won, the _child-rekey_ event is called twice. That the "old" SA is the same in both is definitely not ideal. But the problem is that the "new" SA from the first event might already be destroyed once we trigger the second event (the DELETE for the redundant CHILD_SA the peer created could arrive before the response to the CREATE_CHILD_SA exchange that created the winning SA). I'll try to have a look at this soon.

But why increment/decrement anything during a rekeying? The eventual number of CHILD_SAs doesn't change (that's exactly why _child-updown_ isn't called during rekeyings).

> The second question is, in case of CHILD_REKEY collisions, we sometimes see CHILD_DOWN and other times we do not.

That shouldn't be the case. Please provide more information (in particular what strongSwan version you are using).

**Reply by amvarunn** [reply]
> That shouldn't be the case. Please provide more information (in particular what strongSwan version you are using).

Thank you.
We are using version 5.9.1

**Reply by tobiasbrunner** [reply]
Are you using that version on both sides? Please describe when exactly you see _child-udown_ events during rekeyings. There are unit tests for various collision scenarios (see [here](https://github.com/strongswan/strongswan/blob/master/src/libcharon/tests/suites/test_child_rekey.c)) that try to ensure that the event is not called.

**Reply by tobiasbrunner** [reply]
> I'll try to have a look at this soon.

I've pushed a change to the _ikev2-qske-multi-ke_ branch that refactors the CHILD_SA rekeying (and delete) code so that the _child-rekey_ event is only triggered once in the most common scenarios, in particular for the winner of a rekey collision. Note that I currently don't plan to port that change back to our current master branch because the rekeying was already significantly refactored in the mentioned branch due to support for multiple key exchanges.
